kv/kvserver/concurrency: TestConcurrencyManagerBasic failed

kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13987774?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13987774?buildTab=artifacts#/) on master @ [c5e63c3194932122668db16c4ae2f93fc90191ed](https://github.com/cockroachdb/cockroach/commits/c5e63c3194932122668db16c4ae2f93fc90191ed):


```
        [6] sequence req16: conflicted with ‹00000002-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [6] sequence req16: acquiring latches
        [6] sequence req16: scanning lock table for conflicting locks
        [6] sequence req16: sequencing complete, returned guard
        [7] sequence req17: resolving intent ‹"a"› for txn 00000002 with ABORTED status
        [7] sequence req17: lock wait-queue event: wait for (distinguished) txn 00000003 running request @ key ‹"a"› (queuedLockingRequests: 3, queuedReaders: 0)
        [7] sequence req17: conflicted with ‹00000002-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [7] sequence req17: pushing after 0s for: liveness detection = false, deadlock detection = true, timeout enforcement = false, priority enforcement = false, wait policy error = false
        [7] sequence req17: pushing txn 00000003 to detect request deadlock
        [7] sequence req17: blocked on select in concurrency_test.(*cluster).PushTransaction
        [8] sequence req18: lock wait-queue event: wait for txn 00000003 running request @ key ‹"a"› (queuedLockingRequests: 3, queuedReaders: 0)
        [8] sequence req18: conflicted with ‹00000002-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [8] sequence req18: not pushing
        [8] sequence req18: blocked on select in concurrency.(*lockTableWaiterImpl).WaitOn
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/6586/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/concurrency/concurrency_test_/concurrency_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/shared_locks:599:
        finish [1 args]
        <no input to command>
        ----
        [-] finish req16: finishing request
        [7] sequence req17: lock wait-queue event: done waiting
        [7] sequence req17: conflicted with ‹00000003-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [7] sequence req17: acquiring latches
        [7] sequence req17: scanning lock table for conflicting locks
        [7] sequence req17: sequencing complete, returned guard
        [8] sequence req18: lock wait-queue event: wait for (distinguished) txn 00000004 running request @ key ‹"a"› (queuedLockingRequests: 2, queuedReaders: 0)
        [8] sequence req18: conflicted with ‹00000003-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [8] sequence req18: not pushing
        [8] sequence req18: blocked on select in concurrency.(*lockTableWaiterImpl).WaitOn
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/6586/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/concurrency/concurrency_test_/concurrency_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/shared_locks:612:
        finish [1 args]
        <no input to command>
        ----
        [-] finish req17: finishing request
        [8] sequence req18: lock wait-queue event: done waiting
        [8] sequence req18: conflicted with ‹00000004-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
        [8] sequence req18: acquiring latches
        [8] sequence req18: scanning lock table for conflicting locks
        [8] sequence req18: sequencing complete, returned guard
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/6586/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/concurrency/concurrency_test_/concurrency_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/shared_locks:621:
         
        expected:
        [-] finish req18: finishing request
        
        found:
        [-] finish req18: finishing request
        [-] finish req18: Put [‹"a"›] has held latch for 3957890271 ns. Some possible causes are slow disk reads, slow raft replication, and expensive request processing.
    --- FAIL: TestConcurrencyManagerBasic/shared_locks (8.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36065

--- a/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
@@ -15,6 +15,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"math"
 	"reflect"
 	"regexp"
 	"runtime"
@@ -33,6 +34,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/intentresolver"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/lockspanset"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanlatch"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/txnwait"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -712,6 +714,9 @@ func newCluster() *cluster {
 }
 
 func newClusterWithSettings(st *clustersettings.Settings) *cluster {
+	// Set the latch manager's long latch threshold to infinity to disable
+	// logging, which could cause a test to erroneously fail.
+	spanlatch.LongLatchHoldThreshold.Override(context.Background(), &st.SV, math.MaxInt64)
 	manual := timeutil.NewManualTime(timeutil.Unix(123, 0))
 	return &cluster{
 		nodeDesc:  &roachpb.NodeDescriptor{NodeID: 1},

--- a/pkg/kv/kvserver/spanlatch/manager.go
+++ b/pkg/kv/kvserver/spanlatch/manager.go
@@ -12,6 +12,7 @@ package spanlatch
 
 import (
 	"context"
+	"math"
 	"time"
 	"unsafe"
 
@@ -628,25 +629,27 @@ func (m *Manager) Release(ctx context.Context, lg *Guard) {
 	if lg.snap != nil {
 		lg.snap.close()
 	}
+	m.remove(lg)
 
-	m.mu.Lock()
-	m.removeLocked(lg)
-	m.mu.Unlock()
-	held := timeutil.Now().UnixNano() - lg.acquireTime
-	if lg.acquireTime != 0 && m.settings != nil && held > longLatchHoldDuration.Get(&m.settings.SV).Nanoseconds() {
-		const longLatchHeldMsg = "%s has held latch for %d ns. Some possible causes are " +
+	var held time.Duration
+	if lg.acquireTime != 0 {
+		held = timeutil.Since(timeutil.FromUnixNanos(lg.acquireTime))
+	}
+	if held > m.longLatchHoldThreshold() {
+		const msg = "%s has held latch for %d ns. Some possible causes are " +
 			"slow disk reads, slow raft replication, and expensive request processing."
 		if m.everySecondLogger.ShouldLog() {
-			log.Warningf(ctx, longLatchHeldMsg, lg.baFmt, held)
+			log.Warningf(ctx, msg, lg.baFmt, held)
 		} else {
-			log.VEventf(ctx, 2, longLatchHeldMsg, lg.baFmt, held)
+			log.VEventf(ctx, 2, msg, lg.baFmt, held)
 		}
 	}
 }
 
-// removeLocked removes the latches owned by the provided Guard from the
-// Manager. Must be called with mu held.
-func (m *Manager) removeLocked(lg *Guard) {
+// remove removes the latches owned by the provided Guard from the Manager.
+func (m *Manager) remove(lg *Guard) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 	for s := spanset.SpanScope(0); s < spanset.NumSpanScope; s++ {
 		sm := &m.scopes[s]
 		for a := spanset.SpanAccess(0); a < spanset.NumSpanAccess; a++ {
@@ -661,6 +664,14 @@ func (m *Manager) removeLocked(lg *Guard) {
 			}
 		}
 	}
+}
+
+// longLatchHoldThreshold returns the threshold for logging long latch holds.
+func (m *Manager) longLatchHoldThreshold() time.Duration {
+	if m.settings == nil {
+		return math.MaxInt64 // disable
+	}
+	return LongLatchHoldThreshold.Get(&m.settings.SV)
 }
 
 // Metrics holds information about the state of a Manager.

--- a/pkg/kv/kvserver/spanlatch/settings.go
+++ b/pkg/kv/kvserver/spanlatch/settings.go
@@ -20,10 +20,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 )
 
-var longLatchHoldDuration = settings.RegisterDurationSetting(
+// LongLatchHoldThreshold controls when we will log latch holds.
+var LongLatchHoldThreshold = settings.RegisterDurationSetting(
 	settings.SystemOnly,
 	"kv.concurrency.long_latch_hold_duration",
-	"duration threshold for logging long latch holding",
+	"the threshold for logging long latch holds",
 	3*time.Second,
 	settings.NonNegativeDuration,
 )

